### PR TITLE
Limits decimal places in amount inputs on the home and send screens

### LINKED ISSUE #70: Limit send amount input length to maximum supported by protocol
We need to limit amount input to 10 decimal places at all input areas since that is the most the protocol supports. (home screen, and send flow screen).

### PR DESCRIPTION
**Pull request by CrystalPony**
Fixes #70
I've extended TextInputEditText and added a function that listens for text changes and works like this:

1. The value before text edit is cached
2. If the first character is . or , (to account for currency conversions that use the user's regional format) it adds a leading 0
3. If there are more decimal places than the max argument passed to the function (I set it to 8 to match similar behaviors elsewhere in the app) it compares the current and cached values
- If the current value differs by one symbol we simply restore the cached version - this is to account for situations when the user entered a number somewhere in the middle. If we were to remove the last character the decimal numbers would shift to the right, which looks confusing. Instead we want the value to remain as it were.
- If the current value differs by more than one symbol we remove the last character. This happens if the user copypasted the amount from somewhere, so we want to simply trim it.
4. The leading zeroes are trimmed (we don't want the amount to look like 0001 or 010.998)
5. The cursor position is restored similarly to #175 

For the home screen I've added a new condition to the flow scan of the _typedChars broadcast channel. It checks if there's a dot, and if we already have 8 decimal places it returns the accumulator untouched.

### CODE REVIEW

**Comment by CrystalPony** [comment]
Additional enhancement would be to make a custom currency format so it's uniform across the app - right now sometimes it's '.', sometimes it's ',' and sometimes it adds spaces to separate thousands, however this is minor and out of scope for this task.

File: app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+inline fun TextInputEditText.limitDecimalPlaces(max: Int) {
+    this.apply {
+        var previousValue = ""
+        doBeforeTextChanged { text: CharSequence?, _: Int, _: Int, _: Int ->
+            previousValue = text.toString()
+        }
+        doAfterTextChanged {
+            var textStr = text.toString()
+            if (textStr.contains(Regex("[.,]"))) {
```
Comment (gmale) [inline app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt:17]: I'm a little concerned about the regex use here. Since this happens anytime the text changes, it might make sense to either compile this regex first and reuse it or to split up the logic to handle commas one way and decimals another and simply use `contains(char)`.

File: app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                        if (oneChange) {
+                            textStr = previousValue
+                        } else {
+                            textStr = textStr.removeRange(dotIndex + max + 1, textStr.length)
+                        }
+                    }
+                }
+            }
+
+            val textView = this.findViewById<EditText>(R.id.input_zcash_amount)
```
Comment (gmale) [inline app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt:47]: It might be good to avoid hard coding a layout id in an extension function. Perhaps we can use a more generic approach for extracting the edittext from the input layout, such as the `getEditText` function of that class.

File: app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            val oldText = textView.text.toString()
+
+            textStr = textStr.replaceFirst("^0+(?!$)(?![.,])".toRegex(), "")
```
Comment (gmale) [inline app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt:50]: We can probably achieve this text change without the negative loookaheads. Not everyone understand how they work and it might be difficult to maintain.

File: app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt (hunk 1 above)
Comment (CrystalPony) [reply app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt:17]: Agreed, `contains `would be better here.

File: app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt (hunk 2 above)
Comment (CrystalPony) [reply app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt:47]: The textView var and findViewById are completely redundant, an oversight on my part. Simply using "this" suffices everywhere here.

**Comment by CrystalPony** [comment]
I've commited the changes, also fixed a rare crash related to textStr not always being equal to this.text.toString().

Regarding using CurrencyFormatter, that would be ideal, however that is in the sdk repo. I might take a look at that tomorrow, but I won't modify anything there without consultation. In theory simply creating a custom currency format would suffice, but existing projects might already depend on the current implementation.

**Comment by gmale** [comment]
The changes make sense. For completeness, I want to pull this and test later today because I vaguely remember running into bugs making a similar change to the home screen input. With a bit of testing, I should be able to jog my memory...

**Comment by gmale** [comment]
I couldn't get to testing this yesterday, largely because someone transferred all the funds out of the "developer" wallet so I need to resync a couple other wallets to find my test funds. Hopefully, I'll get this tested today . . .

**Comment by gmale** [comment]
Finally got a bit of time to test. Found a reproducible issue

1. Enter 1234.12345678 on the home screen
2. Press "Send"
3. Type 9 into the amount field
4. Observe: last 4 digits are deleted and then no further typing is permitted
5. Expected: value does not change but no further digits permitted (optionally, with a visual indicator as to why)

And a few other concerns. I'll add them as comments but at this point, I think I would recommend splitting this into two issues, one for the homescreen change, which I think can be merged as is and the other for the send screen, which is a little more involved.

File: app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import cash.z.ecc.android.R
+import com.google.android.material.textfield.TextInputEditText
+
+inline fun TextInputEditText.limitDecimalPlaces(max: Int) {
+    this.apply {
+        var previousValue = ""
+        doBeforeTextChanged { text: CharSequence?, _: Int, _: Int, _: Int ->
+            previousValue = text.toString()
+        }
+        doAfterTextChanged {
```
Comment (gmale) [inline app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt:15]: Using the `doAfterTextChanged` and `doBeforeTextChanged` extension functions in this way, in conjunction with a `previousValue` mutable variable results in two instances of TextWatcher being created. We could perhaps manage state a bit better by directly adding a TextWatcher, rather than relying on the convenience extension functions. For example:

```kotlin

        addTextChangedListener(object : TextWatcher {
            val previousValue = ""
            override fun afterTextChanged(s: Editable?) {
                // move after stuff here
            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
                // move before stuff here
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            }

        })
```

File: app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt [hunk 5]
```diff
@@ -0,0 +1,66 @@
+package cash.z.ecc.android.ext
+
+import android.widget.EditText
+import androidx.core.widget.doAfterTextChanged
+import androidx.core.widget.doBeforeTextChanged
+import cash.z.ecc.android.R
+import com.google.android.material.textfield.TextInputEditText
+
+inline fun TextInputEditText.limitDecimalPlaces(max: Int) {
+    this.apply {
```
Comment (gmale) [inline app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt:10]: with this being an extension function, we could remove the `this.apply` here because in this scope `this` already refers to the TextInputEditText.

File: app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt [hunk 6]
```diff
@@ -0,0 +1,66 @@
+package cash.z.ecc.android.ext
+
+import android.widget.EditText
+import androidx.core.widget.doAfterTextChanged
+import androidx.core.widget.doBeforeTextChanged
+import cash.z.ecc.android.R
+import com.google.android.material.textfield.TextInputEditText
+
+inline fun TextInputEditText.limitDecimalPlaces(max: Int) {
```
Comment (gmale) [inline app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt:9]: This extension function could probably be generalized a bit more and applied to `EditText` widgets, rather than being specific to `TextInputEditText` since it doesn't seem to require anything that is specific to the subclass. If that's the case, then this function could be moved to the existing `cash.z.ecc.android.ext.EditText` file.

File: app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        doBeforeTextChanged { text: CharSequence?, _: Int, _: Int, _: Int ->
+            previousValue = text.toString()
+        }
+        doAfterTextChanged {
+            var textStr = text.toString()
+            if (textStr.contains('.') || textStr.contains(',')) {
+                if (textStr.first() == '.' || textStr.first() == ',') {
+                    textStr = "0$textStr"
+                }
+                if (textStr.last() != '.' && textStr.last() != ',') {
```
Comment (gmale) [inline app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt:21]: It seems like a large portion of this section could be simplified into rounding, perhaps using the SDK conversion functions:

```kotlin
val number = textStr.safelyConvertToBigDecimal();
// only modify the text if it has gotten too long
if (number != null && number.scale() > 8) {
    textStr = number.toZecString(8) // round to 8 places
}
```

**Comment by gmale** [comment]
Quick emulator recording of the behavior, [per request](https://github.com/zcash/zcash-android-wallet/pull/177#issuecomment-652592813):

bug

**Comment by gmale** [comment]
> Thanks for checking this out.
> 
> Weirdly enough I am not able to reproduce this issue. Maybe this is somehow related to regional formats as well.
> If you could make a video or screenshots of the send page before and after you type 9 I'll see if I can figure out what might be causing this.

Added a quick video above (I'm currently in a conference call so it's probably not the best) but note that the first thing I did when I got to the send screen was typed `9`. That's what resulted in `5678` being deleted.

In most other cases, it seems to work as intended but in this repeatable case, it mis-steps and I think it might be related to the `previousValue` being initialized with "" rather than the current value of the textfield. But I'm not sure, I haven't debugged it (because I'm on a 2-hour call :( ).

File: app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt (hunk 7 above)
Comment (CrystalPony) [reply app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt:21]: I tried this approach, however toZecString uses currency formatter which uses the user's locale so I scrapped this approach as I wanted the value here to be formatted uniformly for all users. I didn't realize that this also happens as well when the value is passed from the home screen. More about this in my next comment.

**Comment by CrystalPony** [comment]
Ah, this makes sense. Your locale uses commas to separate thousands, while my locale uses commas to separate decimal places. The current implementation considers both commas and dots as decimal separators as I didn't consider locales like yours.

An easy solution would be to change the way the value is passed from the home screen, but I think that the best approach would be to change the SDK's CurrencyFormatter so that it uses a defined format - this way every user has an identical experience, and edge cases such as this wouldn't be possible.

**Comment by CrystalPony** [comment]
The changes that you proposed regarding using toZecString should solve this as well, however I think that a uniform CurrencyFormatter is a better approach.

**Comment by gmale** [comment]
Sounds about right. The support for locales is awful in this app. I'm sure there are lots of places where I'm doing silly things and mishandle commas v. periods. Let me know if you see any crazy decimal separator problems. I'm surprised it even works in other locales :laughing:

**Comment by gmale** [comment]
> a uniform CurrencyFormatter is a better approach
Agreed, we can make changes there and I can push new versions, if needed. In the short term, I'd recommend writing the functions we want somewhere in the `cash.z.ecc.android.ext` package with the intention of directly moving them over to CurrencyFormmater once they're working as desired.

**Comment by CrystalPony** [comment]
As far as splitting the tasks - I think the gitcoin entries are more of a formality than anything, and I trust you guys, so I don't mind either way. From the development point of view it probably makes sense though.

**Comment by CrystalPony** [comment]
previousValue being initialized with "" shouldn't cause problems as doBeforeTextChanged should always get called before it gets used. If it doesn't get called, such as when the value is set from code, then doAfterTextChanged won't be called as well!

**Comment by CrystalPony** [comment]
> > a uniform CurrencyFormatter is a better approach
> 
> Agreed, we can make changes in the SDK and I can publish new versions, if needed. In the short term, I'd recommend writing the functions we want somewhere in the `cash.z.ecc.android.ext` package with the intention of directly moving them over to CurrencyFormmater once they're working as desired.

Makes sense! I'll start working on the update.

File: app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt (hunk 7 above)
Comment (gmale) [reply app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt:21]: I think we want the numbers formatted according to the user's locale? What do you see as the benefit of doing it uniformly for all?

**Comment by gmale** [comment]
> this way every user has an identical experience

Wait, are you envisioning everyone seeing commas and decimals in the same format? I just had an interesting discussion about this in another thread and was pointed here:
https://en.wikipedia.org/wiki/Decimal_separator#Digit_grouping

TL;DR: there has been a lot of recent standardization around using the `thin space` character as a separator. If we want to go with one standard, then it might be worthwhile to take that into account.

Of course, all of this might be outside of the scope of this issue. For now, we could focus on what works and then make it better, later.

File: app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt (hunk 7 above)
Comment (CrystalPony) [reply app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt:21]: I see quite a few benefits:

1. Someone shared a screenshot of their transaction, which reads "1,234". How do you know if this is 1234 zcash or 1.234 zcash?
2. The home screen currently just uses dots to separate the decimal places. Then when you go to the send screen the format changes to the user's locale, so let's again say 1234 -> 1,234. In general, I think that the format should be the same everywhere in the app, otherwise it looks like a bug. The user might not understand why the number suddenly changes. They might even think that the app added a decimal separator, for example.
What if we apply the user's locale to the home screen as well? This could also get confusing: why does the number format change as the user is typing it, where does the comma come from and what does it mean? I know I might do a double take if I see my 1234 change to 1,234.
The home screen input would also have to be changed to account for this -> if the user's locale uses dots it would have to display a dot key, if it uses commas - then a comma key. This could perhaps introduce new bugs.
3. I've seen the dot separator used pretty much everywhere when it comes to crypto, never seen commas used for separating thousands or the decimal places. I think it makes sense adhere to this emerged standard so to speak.

**Comment by CrystalPony** [comment]
I've implemented all of the requested changes, extended CurrencyFormatter with uniform functionality, and greatly simplified the code.
I've tested this extensively with typing at the end, typing in front, pasting at the end, in front or in the middle and replacing the whole number. Everything seems to work as expected!

**Comment by gmale** [comment]
Reviewed changes. This is excellent. Testing on device, next...

File: app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt (hunk 7 above)
Comment (gmale) [reply app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt]: Good arguments that I would summarize as: we either need to go all-in on localization or uniformly apply one number format, instead. 

Doing localization correctly has benefits. Alternatively, sticking to one format has benefits. Doing **_neither_** and being somewhere in the middle, is the worst option--and that's what we had before. Your changes fix that! :tada:

Comment (gmale) [reply app/src/main/java/cash/z/ecc/android/ext/TextInputEditText.kt]: > pretty much everywhere when it comes to crypto

Side note for context: We try not to use other crypto wallets as reference for this app because, to put it mildly, many crypto wallets are lacking in design. One explicit goal for this app is for UX and design to be a high priority. For things like this, we tend to look toward cashapp or venmo. As a takeaway, I will try to research how those apps handle commas across locales, particularly on input fields.

**Comment by gmale** [comment]
Test passed. Couldn't break it. Merging. Great work!

While I was at it, I also tested your [suggestions on the other bug](https://github.com/zcash/zcash-android-wallet/issues/106) I'll add details on that Issue.
